### ISSUE TITLE
enable type guards over string literal types

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#6028]
Now that the string literal types #5185 are merged in, the type guards based on them would be an [awesome](https://github.com/Microsoft/TypeScript/issues/1003#issuecomment-61130784) addition:

by @ahejlsberg:
> Type guards could relate the common kind property to the corresponding object type in the union and narrow the type appropriately in the guarded block.

The original issue #1003 got closed with the following comment:

> the remaining part is extending the type guards to support the string literal types. but this should be tracked by a different issue.

A quick search suggests that there isn't a ticket for this yet. So here we go.

**Comment by DanielRosenwasser** [comment]
Thanks for filing @aleksey-bykov!

**Comment by Eyas** [comment]
Is this no longer intended for TypeScript 1.8?

**Comment by DanielRosenwasser** [comment]
I predict that'll be going in for 2.0. We're still working on the core functionality of string literal types, and how we want things to flow throughout the type system. If we do that right, this will be much easier and might enable much more interesting features down the road.

**Comment by Eyas** [comment]
@DanielRosenwasser is there an issue/set of issues to track about the type system work?

**Comment by aleksey-bykov** [comment]
what are the more exciting features?

**Comment by DanielRosenwasser** [comment]
[Yes.](https://github.com/Microsoft/TypeScript/labels/Domain%3A%20Literal%20Types) :wink:

**Comment by DanielRosenwasser** [comment]
@aleksey-bykov basically the idea that you'd be able to not only narrow from strings to string literals, but you'd be able to use them as tags, as discussed in #1003.

**Comment by aleksey-bykov** [comment]
good, that is what this issue is about, I thought there might be more to that

**Comment by shaharmor** [comment]
Not sure if its listed anywhere, but i think this should show an error: (Hopefully in the IDE)

```javascript
// declare
type options = 'option1' | 'option2';

interface A {
   possibleOptions: options;
}

/// use (given class B that implements interface A)
var x = new B();
if (x.possibleOptions === 'option3') {
  // there is no way of getting here is all possible values are either 'option1' or 'option2', should error
}

```

**Comment by DanielRosenwasser** [comment]
Can we say that this has been fixed by #9163?

**Comment by aleksey-bykov** [comment]
yes, thanks
